### PR TITLE
[Launching] Only set project-name if IProject is existing IJavaProject

### PR DESCRIPTION
**Pull request by HannesWell**
Fixes https://github.com/eclipse-m2e/m2e-core/issues/1293

@laeubi do you want to verify it fixes the issue too?

### CODE REVIEW

**Comment by laeubi** [comment]
I trust you :+1:

File: org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java [hunk 1]
```diff
@@ -228,14 +229,15 @@ public IVMInstall getVMInstall(ILaunchConfiguration configuration) throws CoreEx
...
+    Optional<IProject> project = getContainer(pomDirectory).map(IContainer::getProject)
+        .filter(p -> JavaCore.create(p).exists());
     if(project.isPresent()) {
       // Set the project name so that super.getVMInstall() called below, can find the JDT-Compiler JDK 
       ILaunchConfigurationWorkingCopy workingCopy = configuration.getWorkingCopy();
       workingCopy.setAttribute(IJavaLaunchConfigurationConstants.ATTR_PROJECT_NAME, project.get().getName());
       configuration = workingCopy;
     }
-    return super.getVMInstall(configuration);
+    return JavaRuntime.computeVMInstall(configuration);
```
Comment (kwin) [inline org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java:240]: why this change?

Comment (HannesWell) [reply org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java:240]: For less indirection and therefore better readability/understandability of the code.
Or do you know a reason that speaks against it?

Comment (laeubi) [reply org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java:240]: **Just for a future enhancement:**
Maybe we even should not use `JavaRuntime.computeVMInstall(configuration);` but mimic whats done there, e.g. for a maven launch, one could go up in the folder hierarchy and check if the folder maps a project in the workspace and simply use the closest parent project.

Beside that I only wonder why we need to set the project at all, should this not be done automatically somehow before, or was always the eclipse default jre used?

**Review by laeubi — APPROVED** [review]

File: org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java (hunk 1 above)
Comment (HannesWell) [reply org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java]: Agree with the first part. And since there is the possibility for clean ups in general I reverted this part again for now and made the change minimal, to reduce the chance for new bugs.
Will provide another PR for a later release that makes that more clear in general.

> Beside that I only wonder why we need to set the project at all, should this not be done automatically somehow before, or was always the eclipse default jre used?

I don't think this is set automatically since a launch-config need not to be connected to a project per-se.

Comment (laeubi) [reply org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java]: So where is the project then used? If only internally by m2e, one might specify a custom (m2e scoped) property ...

Comment (HannesWell) [reply org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java]: It is used by `JavaRuntime.computeVMInstall`, which is from JDT.

Comment (laeubi) [reply org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java]: Yes, but actually do we really *want* the (compile) project VM to execute the maven build? e.g. most of my builds currently user Java 11 but Tycho requires Java 17 (what is the default VM), so for the future I think we should use the Workspace default VM as default and the enforcer one if it is specified.

I think this would here also be the place where one might want to select a different project and/or Workspace VM for launching maven builds.

So the order should be:

1. Enforcer VM
2. Project configured m2e VM (not implemented yet --> needs a project setting)
3. Workspace default m2e VM (not implemented yet --> needs a workspace preference)
4. Default VM of the Workspace.

Comment (HannesWell) [reply org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java]: We should also consider that you can specify in the workspace preferences to which VM installation an EE is bound and IIRC that is also considers.

If that is not sufficient it could be done as suggested, but I assume if the JVM really matters one can set it in the launch config and re-use. At least in my use cases the VM to that should run the build often matches the sources Java-level.

Therefore personally I don't find it important to have extra workspace and project preferences for that. But if somebody want's to implement it and provide a PR, I'm fine with it too. :)

Comment (laeubi) [reply org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java]: EE has no meaning for maven.

Comment (HannesWell) [reply org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java]: Yes, but isn't a EE extract by the JavaRuntime class when it determines the compiler-level of the project?

Comment (laeubi) [reply org.eclipse.m2e.launching/src/org/eclipse/m2e/internal/launch/MavenLaunchDelegate.java]: Not really, if you go to the settings you can choose a **JRE** based on the EE, so if that changes you don't need to change the config, but you can also choose an explicit one (by ID)
